DKG using resharing API (#68)
dedis/drand: README.md

```
go get -u github.com/dedis/drand
```
## Components

drand has two components:

+ a *daemon* that runs the DKG protocol and the randomness beacon
+ a local *control client* that issues command to the daemon so start the DKG
  protocol, or retrieve informations.

The daemon only listens for *localhost* connection from the control client,i.e.
only server's administrators are able to issue commands to their respective drand nodes.

## Usage

This section explains in details the workflow to have a working group of drand
nodes generate randomness. On a high-level, the workflow looks like this:

> + Generate individual longterm key-pair of drand nodes and then the group file that contains all public keys and other informations

Reviewer: ```suggestion
+ Generate individual longterm key-pairs of drand nodes and then the group file that contains all public keys and other information
```

> + Start each drand daemons

Reviewer: ```suggestion
+ Start each drand daemon
```

> + Instruct to each daemons to start the DKG protocol

Reviewer: ```suggestion
+ Instruct each daemon to start the DKG protocol
```

from which we can then assemble the group configuration file, and finally all
participants run the distributed key generation protocol.

#### Long-Term Key

To generate the long-term key pair `drand_id.{secret,public}` of the drand daemon, execute
```
drand keygen <address>
```
where `<address>` is the address from which your drand daemon is reachable. The
address must be reachable over a TLS connection. In case you need non-secured
channel, you can pass the `--insecure` flag.

#### Group Configuration

All informations regarding a group of drand nodes necessary for drand to function properly are located inside a group.toml configuration file. To run a DKG protocol, one needs to generate this group configuration file from all individual longterm keys generated in the previous step. One can do so with:
```
drand group <pk1> <pk2> ... <pkn>
```
where `<pki>` is the public key file `drand_id.public` of the i-th participant.
The group file is generated in the current directory under `group.toml`.
**NOTE:** At this stage, this group file MUST be distributed to all participants !

##### Randomness Beacon Period

> drand updates the configuration file after the DKG protocol finishes,

Reviewer: ```suggestion
After the DKG protocol finishes, drand updates the configuration file
```

> with the distributed public key and automatically starts running the randomness beacon. By default, a randomness beacon has a period of 1mn,i.e. new randomness is generated every minute. If you wish to change the period, you must include that information **inside** the group configuration file. You can do by appending a flag to the command such as :

Reviewer: ```suggestion
with the distributed public key and automatically starts running the randomness beacon. By default, a randomness beacon has a period of 1mn, i.e., new randomness is generated every minute. If you wish to change the period, you must include that information **inside** the group configuration file. You can do so by appending a flag to the command such as :
```

```
drand group --period 2m <pk1> <pk2> ... <pkn>
```
The period must be parsable by the [time](https://golang.org/pkg/time/#ParseDuration) package.

### Starting drand daemon

There are two ways to run a drand daemon: using TLS or using plain old regular
un-encrypted connections. Drand by default tries to use TLS connections.

> The daemon does not go automatically in background, so you must run it with ` &

Reviewer: ```suggestion
The daemon does not go automatically in background, so you must run it with `&`
```

> ` in your terminal, or within a screen / tmux session, or with the `-d` option enabled for the docker commands.

Reviewer: ```suggestion
in your terminal, or within a screen / tmux session, or with the `-d` option enabled for the docker commands.
```

commands below. You must add the corresponding volumes pointing to your TLS
private key and certificate in case you are using TLS, which you should.

**NOTE**: You can test if your daemon is up and running with a ping command on
the same host as the daemon:
```
drand control ping
```

### Distributed Key Generation

After running all drand daemons, each operator needs to issue a command to start
the DKG, using the group file generated before. One can do so with:
```
drand share <group-file>
```

One of the nodes has to function as the leader to initiate the DKG protocol (no
additional trust assumptions), he can do so with:
```
drand share --leader <group-file>
```

Once running, the leader initiates the distributed key generation protocol to
compute the distributed public key (`dist_key.public`) and the private key
shares (`dist_key.private`) together with the participants specified in
`drand_group.toml`.

**Group File**: Once the DKG phase is done, the group file is updated with the

> newly created distributed public key. That updated group file needed by drand to

Reviewer: ```suggestion
newly created distributed public key. That updated group file is needed by drand to
```

securely contact drand nodes on their public interface to gather private or
public randomness. One can get it via the following: 
```bash
drand show group
``` 
It will print the group file in its regular TOML format. If you want to save it to
a file, append the `--out <file>` flag.

**Distributed Public Key**: More generally, for third party implementation of
randomness beacon verification, one only needs the distributed public key. If

> you are an administrator of a drand node, you can use the control port as the

Reviewer: ```suggestion
you are an administrator of a drand node, you can print the distributed public key
```

> following:

Reviewer: with the following command:
